// @ar.io/sdk/lib/cjs/contract-state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ANT_CONTRACT_FUNCTIONS = exports.DENOMINATIONS = exports.AR_IO_CONTRACT_FUNCTIONS = exports.ioContractInteractions = exports.ioContractWriteInteractions = exports.ioContractReadInteractions = void 0;
exports.ioContractReadInteractions = [
    'gateway',
    'gateways',
    'prescribedObservers',
    'record',
    'auction',
    'balance',
    'epoch',
    'priceForInteraction',
];
exports.ioContractWriteInteractions = [
    'delegateStake',
    'decreaseDelegateStake',
    'joinNetwork',
    'increaseOperatorStake',
    'decreaseOperatorStake',
    'updateGatewaySettings',
    'saveObservations',
    'extendRecord',
    'buyRecord',
    'increaseUndernameCount',
    'transfer',
];
exports.ioContractInteractions = [
    ...exports.ioContractReadInteractions,
    ...exports.ioContractWriteInteractions,
];
exports.AR_IO_CONTRACT_FUNCTIONS = {
    GATEWAY: 'gateway',
    GATEWAYS: 'gateways',
    PRESCRIBED_OBSERVERS: 'prescribedObservers',
    DELEGATE_STAKE: 'delegateStake',
    DECREASE_DELEGATE_STAKE: 'decreaseDelegateStake',
    JOIN_NETWORK: 'joinNetwork',
    INCREASE_OPERATOR_STAKE: 'increaseOperatorStake',
    DECREASE_OPERATOR_STAKE: 'decreaseOperatorStake',
    UPDATE_GATEWAY_SETTINGS: 'updateGatewaySettings',
    SAVE_OBSERVATIONS: 'saveObservations',
    EXTEND_RECORD: 'extendRecord',
    INCREASE_UNDERNAME_COUNT: 'increaseUndernameCount',
    RECORD: 'record',
    AUCTION: 'auction',
    TRANSFER: 'transfer',
    BALANCE: 'balance',
    PRICE_FOR_INTERACTION: 'priceForInteraction',
    EPOCH: 'epoch',
    BUY_RECORD: 'buyRecord',
};
// Balances
var DENOMINATIONS;
(function (DENOMINATIONS) {
    DENOMINATIONS["IO"] = "IO";
    DENOMINATIONS["MIO"] = "mIO";
})(DENOMINATIONS || (exports.DENOMINATIONS = DENOMINATIONS = {}));
exports.ANT_CONTRACT_FUNCTIONS = {
    TRANSFER: 'transfer',
    SET_CONTROLLER: 'setController',
    REMOVE_CONTROLLER: 'removeController',
    SET_NAME: 'setName',
    SET_TICKER: 'setTicker',
    SET_RECORD: 'setRecord',
    REMOVE_RECORD: 'removeRecord',
    BALANCE: 'balance',
    EVOLVE: 'evolve',
};
