// @ar.io/sdk/lib/cjs/utils/http-client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAxiosInstance = void 0;
/**
 * Copyright (C) 2022-2024 Permanent Data Solutions, Inc. All Rights Reserved.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
const axios_1 = __importDefault(require("axios"));
const axios_retry_1 = __importDefault(require("axios-retry"));
const logger_js_1 = require("../common/logger.js");
const version_js_1 = require("../version.js");
const createAxiosInstance = ({ axiosConfig = {}, logger = logger_js_1.Logger.default, retryConfig = {
    retries: 5,
    retryDelay: axios_retry_1.default.exponentialDelay,
    retryCondition: (error) => axios_retry_1.default.isRetryableError(error),
    onRetry(retryCount, error, requestConfig) {
        logger.error(`Retrying request ${requestConfig.url} attempt ${retryCount}`, error);
    },
}, } = {}) => {
    const axiosInstance = axios_1.default.create({
        ...axiosConfig,
        maxRedirects: 0,
        headers: {
            ...axiosConfig.headers,
            'x-source-version': `${version_js_1.version}`,
            'x-source-identifier': 'ar-io-sdk',
        },
        validateStatus: () => true, // don't throw on non-200 status codes
    });
    // add retries
    (0, axios_retry_1.default)(axiosInstance, retryConfig);
    return axiosInstance;
};
exports.createAxiosInstance = createAxiosInstance;
