// @ar.io/sdk/lib/esm/utils/http-client.js
/**
 * Copyright (C) 2022-2024 Permanent Data Solutions, Inc. All Rights Reserved.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import axios from 'axios';
import axiosRetry from 'axios-retry';
import { Logger } from '../common/logger.js';
import { version } from '../version.js';
export const createAxiosInstance = ({ axiosConfig = {}, logger = Logger.default, retryConfig = {
    retries: 5,
    retryDelay: axiosRetry.exponentialDelay,
    retryCondition: (error) => axiosRetry.isRetryableError(error),
    onRetry(retryCount, error, requestConfig) {
        logger.error(`Retrying request ${requestConfig.url} attempt ${retryCount}`, error);
    },
}, } = {}) => {
    const axiosInstance = axios.create({
        ...axiosConfig,
        maxRedirects: 0,
        headers: {
            ...axiosConfig.headers,
            'x-source-version': `${version}`,
            'x-source-identifier': 'ar-io-sdk',
        },
        validateStatus: () => true, // don't throw on non-200 status codes
    });
    // add retries
    axiosRetry(axiosInstance, retryConfig);
    return axiosInstance;
};
